Added gender neutral rewrite transformation

tonysun9 opened the pull request.

kaustubhdhole commented:
Pinging @ashish3586 @Saad-Mahamood

Saad-Mahamood commented on transformations/gender_neutral_rewrite/transformation.py:
```diff
@@ -0,0 +1,425 @@
+    # if (1) input is longer (e.g. document) or (2) we wish to have faster computation, we can set longer stride
+    # reference: https://huggingface.co/transformers/perplexity.html
+
+    if stride == 1:  # if stride is 1, just return average log prob of each token (no need to copy and mask)
```
One question. Given the use case of here is at a sentence level. Will there ever be a need to have a bigger `stride` value? If not is there any need for the code underneath this `if` statement?

tonysun9 replied:
Good observation. Generally, setting `stride` to be 1 is appropriate. However, in certain situations, a larger value may be useful. For example, when compute power is an issue, setting `stride` to be 2, 3, or even 5 can help immensely. From my experience playing around with this parameter, having larger values can significantly improve runtime without noticeable decrease in the quality of rewrites.

Saad-Mahamood reviewed (changes requested):
Generally well written and documented code. I only have one improvement request and one query before I can approve.

kaustubhdhole commented:
@tonysun9 would you like to address the above comments ?

tonysun9 commented on transformations/gender_neutral_rewrite/myconstants.py:
```diff
@@ -0,0 +1,66 @@
+# pronouns
```
Thanks for the suggestion. We certainly would encourage the addition of more occupational words and gender-neutral inflections. We chose to use a `.py` file for the ability to add comments where necessary, since it may not always be immediately obvious why a certain mapping is included. We hope that users can easily add gender or occupation-specific words by following the existing template (perhaps adding an instruction to the README may not be a bad idea as well).

kaustubhdhole commented on transformations/gender_neutral_rewrite/transformation.py:
```diff
@@ -0,0 +1,425 @@
+# https://spacy.io/models/en
+import en_core_web_sm
+
+nlp = en_core_web_sm.load()
```
Can you load this using the global spacy load?
```python
from initialize import spacy_nlp
import spacy
self.nlp = spacy_nlp if spacy_nlp else spacy.load("en_core_web_sm")
```

tonysun9 replied:
Done. Thanks for the suggestion.

kaustubhdhole commented:
@Saad-Mahamood @ashish3586

tonysun9 commented on transformations/gender_neutral_rewrite/transformation.py:
```diff
@@ -0,0 +1,425 @@
+    ]
+    languages = ["en"]
+
+    def __init__(self, seed=0, max_outputs=1):
```
Done. Thanks for the suggestion.

tonysun9 commented:
@kaustubhdhole @Saad-Mahamood I believe I've addressed the comments appropriately. Could you please take a look when you get a chance?

Saad-Mahamood commented on transformations/gender_neutral_rewrite/transformation.py:
```diff
@@ -0,0 +1,482 @@
+        TaskType.TEXT_TAGGING,
+    ]
+    languages = ["en"]
+
```
Missing the [keywords](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md) variable here.

tonysun9 replied:
Added.

Saad-Mahamood reviewed (changes requested):
Missing keywords variable in transform.py needed before this can be merged.

ashish3586 commented on transformations/gender_neutral_rewrite/README.md:
```diff
@@ -0,0 +1,127 @@
+# Gender Neutral Translation
+This perturbation adds noise to all types of text sources (sentence, paragraph, etc.) proportional to noise erupting 
```
The "Gender Neutral Translation" definition seems different from the implementation.

tonysun9 replied:
Thanks for the catch. I've removed these lines in the README.

ashish3586 commented on transformations/gender_neutral_rewrite/transformation.py:
```diff
@@ -0,0 +1,493 @@
+    return new_sentence
+
+
+class GenderNeutralRewrite(SentenceOperation):
```
Please add docstring for the GenderNeutralRewrite class.

tonysun9 replied:
Done. Thanks for the suggestion.

tonysun9 commented:
> @kaustubhdhole @Saad-Mahamood I believe I've addressed the comments appropriately. Could you please take a look when you get a chance?

Would it be possible to get an extension on the 09/30 deadline? I believe I have addressed all the comments appropriately.

kaustubhdhole commented:
@tonysun9 please fix the test cases. Will merge then.

tonysun9 commented:
> @tonysun9 please fix the test cases. Will merge then.

I think the failing test case is not related to my pull request?

kaustubhdhole commented:
Seems that is the same error as that in main. Okay, I will merge this then. One last thing, please move the spacy loading into your constructor rather than in the import file. Thanks for your hard work and thanks @Saad-Mahamood and @ashish3586 for reviewing the code.

tonysun9 commented:
> Seems that is the same error as that in main. Okay, I will merge this then. One last thing, please move the spacy loading into your constructor rather than in the import file. Thanks for your hard work and thanks @Saad-Mahamood and @ashish3586 for reviewing the code.

Done. Thanks for the suggestion and taking the time to review the code.
